### PR TITLE
Add Crystal::System::Dir

### PR DESCRIPTION
**Pull request by RX14**
I decided to expose `Crystal::System::Dir` as an `Iterator(String)` since I think it's the nicest yet most generic interface for what `readdir` actually is: an iterator.

### CODE REVIEW

**Comment by larubujo** [comment]
seem dir missing finalize, maybe from before. so iterator should have finalize. so class, not struct.

**Comment by RX14** [comment]
@larubujo yes, `Dir` is missing a `def finalize`, thanks for spotting.

File: src/dir.cr [hunk 1]
```diff
@@ -17,10 +15,7 @@ class Dir
 
   # Returns a new directory object for the named directory.
   def initialize(@path)
-    @dir = LibC.opendir(@path.check_no_null_byte)
-    unless @dir
-      raise Errno.new("Error opening directory '#{@path}'")
-    end
+    @dir = Crystal::System::Dir.each(@path)
```
Comment (straight-shoota) [inline src/dir.cr]: This ivar should perhaps be renamed to match the new meaning? For example `@entry_iterator`.

File: src/dir.cr [hunk 2]
```diff
@@ -101,7 +96,7 @@ class Dir
   end
 
   def each_child
-    ChildIterator.new(self)
+    @dir.reject { |x| {".", ".."}.includes?(x) }
```
Comment (straight-shoota) [inline src/dir.cr]: I'm not sure if/why this is preferable over a custom iterator class?

File: src/dir.cr [hunk 3]
```diff
@@ -65,7 +60,7 @@ class Dir
   end
 
   def each_entry
-    EntryIterator.new(self)
+    @dir
```
Comment (lbguilherme) [inline src/dir.cr]: `each_entry` should return a new iterator each it is called, no?

Comment (RX14) [reply src/dir.cr]: If you look at the behaviour of the old iterator, this wasn't how it worked. The whole `Dir` class is a very poor iterator, the `each_entry` iterator simply wrapped it up into a real `Iterator`, but the state was still in `Dir`. This, while confusing, keeps the old `Dir` semantics.

Comment (RX14) [reply src/dir.cr]: I plan to clean up `Dir` after merging this PR, and make it a module with entirely class methods.

Comment (lbguilherme) [reply src/dir.cr]: I disagree with the old semantics, but I do agree that this PR should not change the semantics. So let it be.

Comment (RX14) [reply src/dir.cr]: I think everyone sane disagrees with the old semantics!

File: src/dir.cr (hunk 2 above)
Comment (RX14) [reply src/dir.cr]: It's easier to read, and a custom iterator would have to wrap `@dir` regardless. So comparing this to a custom iterator, the only overhead would be a single function pointer call overhead. Given that every call to `next` will perform a syscall (which busts cache and takes 200+ cycles!): this is a very acceptable overhead.

File: src/dir.cr (hunk 1 above)
Comment (RX14) [reply src/dir.cr]: Fixed

**Comment by RX14** [comment]
Rebased to fix travis.

**Comment by RX14** [comment]
Replaced this with a more "traditional" port to `Crystal::System`, and I'll refactor `Dir` later.

**Comment by RX14** [comment]
Interesting bug:
```
S-tring.o: In function `*String#<=><String>:Int32':
String:(.text+0x4022): undefined reference to `*Int32@Number#<=><Int32>:Int32'
String:(.text+0x4038): undefined reference to `*Int32@Number#sign:Int32'
/usr/bin/ld: /tmp/crystal/crystal-run-crystal-spec-output.tmp: No symbol version section for versioned symbol `*Int32@Number#<=><Int32>:Int32'
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
Error: execution of command failed with code: 1: `cc "${@}" -o '/tmp/crystal/crystal-run-crystal-spec-output.tmp'  -rdynamic  -lpcre -lgc -lpthread /mnt/src/ext/libcrystal.a -levent -lrt -ldl -L/usr/lib -L/usr/local/lib`
```

re running to see if it's intermittent.

**Comment by RX14** [comment]
I've only seen the above error once before, and it doesn't seem to be reproducible. Perhaps a rare race condition in the cache?

**Comment by asterite** [comment]
I got race conditions in the cache many times. It would be nice to have an issue to track this, with possible suggestions for solutions.

**Comment by RX14** [comment]
@asterite a key

**Comment by RX14** [comment]
@asterite this should really be another issue, but we should codegen into tempoary files `$cache/temp/random.o` then perform a move operation into the final destination once we know the the compilation works (i.e. linking finishes). Moving files on the same FS is an atomic operation. That fixes the case of an interrupted (ctrl-c) compilation corrupting the cache.

This specific crash i've never ever seen any error message like it, and it can't be related to an interrupted compiler, so that's very strange.

**Pull request merged**